3rdparty/mbedtls/CMakeLists.txt

```diff
@@ -1,146 +1,187 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-# NOTE: Because we still build mbed TLS via `ExternalProject_Add`, it
-# does not automatically inherit our build settings (nor could it
-# because it doesn't link against our targets). Unfortunately, that
-# means we have to maintain these flags to be consistent with
-# those used to build `oecore`.
-list(APPEND MBEDTLS_WRAP_CFLAGS_LIST
-  ${OE_SPECTRE_MITIGATION_FLAGS}
-  -nostdinc
-  -I${PROJECT_SOURCE_DIR}/include/openenclave/corelibc
-  -I${PROJECT_SOURCE_DIR}/include
-  -DOE_NEED_STDC_NAMES
-  -fPIE
-  -ftls-model=local-exec
-  -fvisibility=hidden
-  -fno-stack-protector
-  -fno-builtin-udivti3)
-
+# Load the compiler info if we're cross compiling on Windows.
 if (USE_CLANGW)
-  list(APPEND MBEDTLS_WRAP_CFLAGS_LIST "-target x86_64-pc-linux")
-  # Setup clang wrapper toolchain
-  set(MBEDTLS_TOOLCHAIN "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_CURRENT_SOURCE_DIR}/toolchain-clangw.cmake")
-
-  # Setup linker
-  find_program(LLVM_AR "llvm-ar.exe")
-  set(MBEDTLS_LINKER "-DCMAKE_AR=${LLVM_AR}")
-  set(MBEDTLS_COMPILER "")
-  set(CMAKE_AR ${LLVM_AR})
-else()
-  set(MBEDTLS_TOOLCHAIN "")
-  set(MBEDTLS_LINKER "")
-  set(MBEDTLS_COMPILER "-DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}")
+  include(toolchain-clangw.cmake)
 endif()
 
-# Create a patched version of mbed TLS's `config.h` that the external
-# project depends on.
+# We use configure_file since the config.h file has #cmakedefines in it, so
+# the C header changes if we're running in debug mode or not. This will copy
+# the config file to the build directory.
 string(TOUPPER ${CMAKE_BUILD_TYPE} CMAKE_BUILD_TYPE)
 if ("${CMAKE_BUILD_TYPE}" STREQUAL "DEBUG")
   set(MBEDTLS_CERTS_C TRUE)
   set(MBEDTLS_DEBUG_C TRUE)
 endif ()
+configure_file(config.h config.h)
+
+# Compiler flags is copied from the mbedtls CMakeLists.txt and
+# mbedtls/library/CMakeLists.txt files, so that we can compile with the same warnings.
+string(REGEX MATCH "Clang" CMAKE_COMPILER_IS_CLANG "${CMAKE_C_COMPILER_ID}")
+string(REGEX MATCH "GNU" CMAKE_COMPILER_IS_GNU "${CMAKE_C_COMPILER_ID}")
+
+set(MBEDTLS_COMPILE_OPTS
+  -W
+  -Wdeclaration-after-statement
+  -Wwrite-strings
+  -Wshadow
+  # Disable conversion warnings inherited from OE.
+  -Wno-sign-conversion
+  -Wno-conversion)
```

Comment:
Can you confirm if mbedtls is being compiled with 'enclave-appropriate' flags?

- -fPIE  - needed for getting the compiler to generate code optimal for use in an executable/first loaded module. Without this there could be relocations that our loader may not be able to handle.
- -fvisibility=hidden - to keep the enclave self-contained and also helps keep the loader simple
- -fno-stack-protector - We don't support stack protector yet. Needs more work in our C library startup.
- -ftls-model=local-exec - This is the most optimal thread-local storage model - best for use in a single module that is loaded. All TLS variables are accessed via negative offsets off FS register.

Without these flags, there will be loading errors for some enclaves; but common enough, but hard to predictably reproduce.

They ought to get inherited from oelibc dependency; but can you double-check since cmake has surprised us in the past.

Reply:
yup they are all there